### ISSUE TITLE
OOM when implementing training_epoch_end

### ISSUE THREAD

**Issue by WrRan** [Lightning-AI/lightning#15514]
### Bug description

See the toy example:
```python
import pytorch_lightning as pl
import torch
import torch.nn.functional as F
from torch import nn
from torch.utils.data import DataLoader, Dataset


class SomeModel(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.linear = nn.Linear(1, 2)

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=1e-3)

    def forward(self):
        batch_size, features = 10240, 24000
        x_hat = torch.randn(batch_size, features, requires_grad=True, device=self.device)
        x = torch.randn(batch_size, features, device=self.device)
        loss = F.mse_loss(x_hat, x)
        self.log("train_loss", loss)
        return {
            "loss": loss,
            "x": x,
        }

    def training_step(self, batch, batch_idx):
        return self()

    def training_epoch_end(self, outputs) -> None:
        # do nothing
        ...


class SomeDataset(Dataset):
    def __len__(self):
        return 100000000

    def __getitem__(self, index):
        return "some sample"


train = SomeDataset()
val = SomeDataset()

model = SomeModel()
trainer = pl.Trainer(devices=1, accelerator="gpu", check_val_every_n_epoch=1, )
trainer.fit(model, DataLoader(train), DataLoader(val))
```

Run it, and we can get an OOM error.

### How to reproduce the bug

```python
see the above.
```

### Error messages and logs

```
RuntimeError: CUDA out of memory.
```

### Environment

PyTorch Lightning Version : 1.7.7

### More info

How to fix it:
~~the bug comes from the method `advance` of `training_epoch_loop`.~~ **WRONG**

~~see https://github.com/Lightning-AI/lightning/blob/1.7.7/src/pytorch_lightning/loops/epoch/training_epoch_loop.py#L213~~

~~when we save the `batch_end_outputs` for use, we should **detach every tensors in the output**.~~

cc @carmocca @justusschock @rohitgr7

**Comment by carmocca** [comment]
This is expected. When you override `training_epoch_end`, we store all batch outputs for the hook.

We don't want to detach because the user might want the grads.

If you don't want to do this, but want an epoch end hook, use `on_train_epoch_end` which does not have this behaviour

**Comment by awaelchli** [comment]
@carmocca I think it is generally right that the user "could" do something with it. However, it won't be the gradients themselves. The training step has already done the optimization and back-propagating again through all graphs in `training_epoch_end` is not going to work. I don't think it is not the worst idea to detach(). Are there any use cases that I missed?

**Comment by WrRan** [comment]
It seems inconsistent to me because the below code does not throw an OOM exception.
```python
# All parts are the same as the code above, 
# the only difference is that the returned values of `forward` is no longer **a loss plus a detached tensor**.
# ...
    def forward(self):
        batch_size, features = 10240, 24000
        x_hat = torch.randn(batch_size, features, requires_grad=True, device=self.device)
        x = torch.randn(batch_size, features, device=self.device)
        loss = F.mse_loss(x_hat, x)
        self.log("train_loss", loss)
        return {
            "loss": loss,
        }
```

**Comment by WrRan** [comment]
And I agree with @awaelchli. I cannot imagine the use cases where developers keep gradients of the whole dataset in a practical situation.

**Comment by awaelchli** [comment]
If @carmocca agrees, @WrRan would you be interested in sending a PR for this change? We can then let the test suite run and check whether this breaks anything non-obvious that we missed.

**Comment by WrRan** [comment]
@awaelchli It is my pleasure to make a contribution to pl. :)

**Comment by snknitin** [comment]
Use `loss.item()` when you aggregate your losses across batches at the end of the epoch. Especially for logging. Found that after reading through a lot of issues and stackoverflow posts. Compiled some useful suggestions for solving CUDA out of memory issues  [here](https://medium.com/@snk.nitin/how-to-solve-cuda-out-of-memory-error-850bb247cfb2?sk=165a35fded1811fa749756cb1f7b5c19)

**Comment by WrRan** [comment]
@snknitin Thanks for your attention. But the OOM issue does not come from the `self.log("train_loss", loss)`. It comes from the returned value `{"loss": loss, "x": x,}`, and we cannot return `loss.item()` because **pl** needs non-detached loss to optimize models.

**Comment by WrRan** [comment]
Sorry for my wrong solution.

In fact, the `batch_output` and `batch_end_outputs` do not contain any tensors attached in the computing graph.  And when removing the custom empty `training_epoch_end`, we can get rid of the OOM. So the bug is related to the https://github.com/Lightning-AI/lightning/blob/1.7.7/src/pytorch_lightning/loops/epoch/training_epoch_loop.py#L226.
However, the following https://github.com/Lightning-AI/lightning/blob/1.7.7/src/pytorch_lightning/loops/epoch/training_epoch_loop.py#L227 does not hold any attached tensors.
Then, I checked [training_batch_loop.py::advance](https://github.com/Lightning-AI/lightning/blob/1.7.7/src/pytorch_lightning/loops/batch/training_batch_loop.py#L70-L94), [optimizer_loop.py::advance](https://github.com/Lightning-AI/lightning/blob/1.7.7/src/pytorch_lightning/loops/optimization/optimizer_loop.py#L198-L206) and [closure.py::consume_result](https://github.com/Lightning-AI/lightning/blob/1.7.7/src/pytorch_lightning/loops/optimization/closure.py#L44-L57), nothing related to the custom `training_epoch_end` was found.

In short, I have no idea to fix this issue.
Do you have any ideas?
@carmocca @justusschock @rohitgr7 @awaelchli

**Comment by snknitin** [comment]
> @snknitin Thanks for your attention. But the OOM issue does not come from the `self.log("train_loss", loss)`. It comes from the returned value `{"loss": loss, "x": x,}`, and we cannot return `loss.item()` because **pl** needs non-detached loss to optimize models.

So in the return value it obviously needs to be loss for backprop. But in the log, you don't need the whole object. in `self.log("train_loss", loss.item())` can be done

Also in the `return {"loss": loss, "x": x,}` unless you're doing something with x down the line, like metrics calculation, there is probably no need for it to be accumulating across batches on GPU. You might do `x.cpu()` or `x.detach()` . Looks like x is just the labels. This is probably the main culprit taking up most of the memory

**Comment by WrRan** [comment]
@snknitin Using `x.detach`, we will get OOM. Using `x.cpu()`, the training speed is slown down and we will get OOM of CPU memory.

**Comment by WrRan** [comment]
Thanks to @snknitin's hint, the OOM comes from the https://github.com/Lightning-AI/lightning/blob/1.7.7/src/pytorch_lightning/loops/epoch/training_epoch_loop.py#L227. Because the `x` is large (`10240 x 24000 x 8 / 1024^{3} = 1.83`GB), the memory is expected to run out. **Therefore, this issue is a feature instead of a bug**. 

Though, [the inconsistency](https://github.com/Lightning-AI/lightning/issues/15514#issuecomment-1304387684) really confuses me.

Can you @carmocca @justusschock @rohitgr7 @awaelchli have a look?

**Comment by WrRan** [comment]
So what about checking the signature of the overriden `training_epoch_end`? When users do not need the `outputs`, skip the [L227](https://github.com/Lightning-AI/lightning/blob/1.7.7/src/pytorch_lightning/loops/epoch/training_epoch_loop.py#L227).

Actually, I use `training_epoch_end` to do some log as followings:
```python
def training_epoch_end(self, outputs: EPOCH_OUTPUT) -> None:
    self.log("train_acc", self.train_accuracy.compute(), sync_dist=True)
    self.log("train_bleu_1", self.train_bleu_1.compute(), sync_dist=True)
    self.log("train_bleu_2", self.train_bleu_2.compute(), sync_dist=True)
```

**Comment by carmocca** [comment]
You should be using `on_train_epoch_end` if you don't use the outputs.

https://github.com/Lightning-AI/lightning/issues/8731 suggests removing this hook because of the confusion you had

**Comment by WrRan** [comment]
Thanks! 😄
